contributors/design-proposals/pod-priority-api.md

```diff
@@ -0,0 +1,237 @@
+reserved for the system pods and its value is the largest 32-bit
+integer. The cluster may have many more user defined priority classes for
+various use cases. The following list is an example of how the priorities and
+their values may look like.
+We decided to ship Kubernetes with no predefined priority classes, except
+for the "system" priority which will be built in Kubernetes. The main goal of
+of having no built-in priority classes is to avoid creating defacto standard
+names which may be hard to change in the future.
+
+```
+system  2147483647 (int_max)
+tier1   4000
+tier2   2000
+tier3   1000
+```
+
+The following shows a list of example workloads in a Kubernetes cluster in decreasing order of priority:
+
+* Kubernetes system daemons (per-node like fluentd, and cluster-level like 
+  Heapster)
+* Critical user infrastructure (e.g. storage servers, monitoring system like
+  Prometheus, etc.)
+* Components that are in the user-facing request serving path and must be able 
+  to scale up arbitrarily in response to load spikes (web servers, middleware,
+  etc.)
+* Important interruptible workloads that need strong guarantee of
+  schedulability and of not being interrupted
+* Less important interruptible workloads that need a less strong guarantee of
+  schedulability and of not being interrupted
+* Best effort / opportunistic
+
+### Resolving priority class names 
+
+User requests sent to Kubernetes may have "PriorityClassName" in their PodSpec.
+Admission controller resolves a PriorityCalssName to its corresponding number
+and populates the "Priority" field of the pod spec. The rest of Kubernetes
+components look at the "Priority" field of pod status and work with the integer
+value. In other words, "PriorityClassName" will be ignored by the rest of the
+system.
+
+We are going to add a new API object called PriorityClass. The priority class
+defines the mapping between the priority name and its value. It can have an
+optional description. The description is an arbitrary string and is provided
+only as a guideline for users.
+
+If a pod does not specify any "PriorityClassName", the system resolves it to 
+zero in the first version, but we will add support to annotate one of the
+`PriorityClass` objects that tells the system that it is the default one. When
+there is a default priority class, a pod with no "PriorityClassName" will get
+the value of this default class (instead of zero).
+
+```
+type PriorityClass struct {
+  metav1.TypeMeta
+  // +optional
+  metav1.ObjectMeta
+  
+  // The value of this priority class. This is the actual priority that pods
+  // recieve when they have the above name in their pod spec.
+  Value        int32
+  
+  // A description of this priority class, such as the recommended use cases.
+  // +optional
+  Description  string
+}
+```
+
+And the annotation for the default class will be something like:
+
+```yaml
+...
+annotations:
+  priorityclass.alpha.kubernetes.io/is-default-class: "true"
```

Comment:
i think we are trying to avoid bool style fields (even in annotations), in favor of something more policy like.

https://github.com/kubernetes/community/pull/627

so maybe we could just have

`priorityclass.alpha.kubernetes.io/admissionPolicy: "default"`

we would need to say what happens when or if there are multiple defaults.  i assume last modified wins?